FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/pkcs11_operations.c

```diff
@@ -0,0 +1,1038 @@
+    int xMbedResult = 0;
+    mbedtls_pk_context xMbedPkContext = { 0 };
+
+    mbedtls_pk_init( &xMbedPkContext );
+    xMbedResult = mbedtls_pk_parse_key( &xMbedPkContext, ( const uint8_t * ) pcPrivateKey,
+                                        xPrivateKeyLength, NULL, 0 );
+
+    if( xMbedResult != 0 )
+    {
+        LogError( ( "Unable to parse private key." ) );
+        xResult = CKR_ARGUMENTS_BAD;
+    }
+
+    /* Determine whether the key to be imported is RSA or EC. */
+    if( xResult == CKR_OK )
+    {
+        xMbedKeyType = mbedtls_pk_get_type( &xMbedPkContext );
+
+        if( xMbedKeyType == MBEDTLS_PK_RSA )
+        {
+            xResult = prvProvisionPrivateRSAKey( xSession, pcLabel, &xMbedPkContext );
+        }
+        else if( ( xMbedKeyType == MBEDTLS_PK_ECDSA ) ||
+                 ( xMbedKeyType == MBEDTLS_PK_ECKEY ) ||
+                 ( xMbedKeyType == MBEDTLS_PK_ECKEY_DH ) )
+        {
+            xResult = prvProvisionPrivateECKey( xSession, pcLabel, &xMbedPkContext );
+        }
+        else
+        {
+            LogError( ( "Invalid private key type provided. Only RSA-2048 and "
+                        "EC P-256 keys are supported." ) );
+            xResult = CKR_ARGUMENTS_BAD;
+        }
+    }
+
+    mbedtls_pk_free( &xMbedPkContext );
+
+    return xResult;
+}
+
+/*-----------------------------------------------------------*/
+
+static CK_RV prvProvisionCertificate( CK_SESSION_HANDLE xSession,
+                                      const char * pcCertificate,
+                                      size_t xCertificateLength,
+                                      const char * pcLabel )
+{
+    PKCS11_CertificateTemplate_t xCertificateTemplate;
+    CK_OBJECT_CLASS xCertificateClass = CKO_CERTIFICATE;
+    CK_CERTIFICATE_TYPE xCertificateType = CKC_X_509;
+    CK_FUNCTION_LIST_PTR xFunctionList = NULL;
+    CK_RV xResult = CKR_OK;
+    uint8_t * pucDerObject = NULL;
+    int32_t ulConversion = 0;
+    size_t xDerLen = 0;
+    CK_BBOOL xTokenStorage = CK_TRUE;
+    CK_BYTE pxSubject[] = "TestSubject";
+    CK_OBJECT_HANDLE xObjectHandle = CK_INVALID_HANDLE;
+
+    /* Initialize the client certificate template. */
+    xCertificateTemplate.xObjectClass.type = CKA_CLASS;
+    xCertificateTemplate.xObjectClass.pValue = &xCertificateClass;
+    xCertificateTemplate.xObjectClass.ulValueLen = sizeof( xCertificateClass );
+    xCertificateTemplate.xSubject.type = CKA_SUBJECT;
+    xCertificateTemplate.xSubject.pValue = pxSubject;
+    xCertificateTemplate.xSubject.ulValueLen = strlen( ( const char * ) pxSubject );
+    xCertificateTemplate.xValue.type = CKA_VALUE;
+    xCertificateTemplate.xValue.pValue = ( CK_VOID_PTR ) pcCertificate;
+    xCertificateTemplate.xValue.ulValueLen = ( CK_ULONG ) xCertificateLength;
+    xCertificateTemplate.xLabel.type = CKA_LABEL;
+    xCertificateTemplate.xLabel.pValue = ( CK_VOID_PTR ) pcLabel;
+    xCertificateTemplate.xLabel.ulValueLen = strlen( pcLabel );
```

Comment:
```suggestion
    xCertificateTemplate.xLabel.ulValueLen = strnlen( pcLabel, pkcs11configMAX_LABEL_LENGTH );
```

Per the pkcs11 spec, pcLabel should not be null termianted and isn't necessarily a a string, but rather up to 32 bytes of arbitrary data padded at the end with the space ' ' character.

Reply:
Resolved in this commit (along with other labels with similar issues): https://github.com/FreeRTOS/FreeRTOS/pull/724/commits/286f3f5c78e9b0ed419b68e7325f3f02683bc5fc